### ISSUE TITLE
Sending C2D Messages Fails With No Errors

### ISSUE THREAD

**Issue by michaelbironneau** [Azure/azure-iot-sdk-node#385]
# Context
- **OS and version used:** Azure Function (Windows 32-bit)
- **Node.js version:**  6.5.0
- **npm version:** 3.10.0
- **list of installed packages:** 
```
-- azure-iothub@1.7.4
  +-- async@2.6.1
  +-- azure-iot-amqp-base@2.0.0
  | +-- lodash.merge@4.6.1
  | +-- uuid@3.3.2
  | `-- ws@6.0.0
  |   `-- async-limiter@1.0.0
  +-- azure-iot-common@1.7.4
  | `-- getos@3.1.0
  |   `-- async@2.4.0
  +-- azure-iot-http-base@1.7.4
  +-- debug@3.2.5
  | `-- ms@2.1.1
  +-- lodash@4.17.11
  +-- machina@2.0.2
  | `-- lodash@3.10.1
  `-- rhea@0.2.20
```

# Description of the issue:
This may be a duplicate of #353 but our issue is not intermittent, it is constant and reproducible. Triggering the Azure Function that uses the below code, the callback is never called. 

# Code sample exhibiting the issue:
```
var Client = require('azure-iothub').Client;
var Message = require('azure-iot-common').Message;
var NoRetry = require('azure-iot-common').NoRetry;
var connectionString = 'HostName=*****.azure-devices.net;SharedAccessKeyName=iothubowner;SharedAccessKey=*******************';


function validate(msg){
    if (!msg.device_code) throw('No device code for message');
    if (!msg.entities) throw('Entities not specified for message');
    if (!msg.type) throw('Type not specified for message');
    if (!msg.content) throw('Content not specified for message');
}

module.exports = function(context, mySbMsg) {
    var serviceClient = Client.fromConnectionString(connectionString);
     context.log('Received request at ' + new Date());
     serviceClient.setRetryPolicy(new NoRetry());
     
     serviceClient.open(function (err, result) {
         serviceClient.on('disconnect', function(){
             context.log('--Client disconnected--');
         });
         serviceClient.on('connect', function(){
             context.log('--Client connected--');
         });
        context.log(result);
        context.log('Opened service client at ' + new Date());
       if (err) {
         serviceClient.close();
         context.done('Could not connect to IoTHub: ' + err.message); 
         return;
       } else {
         try {
             validate(mySbMsg);
         } catch (ex) {
             context.log('Invalid message', ex);
             serviceClient.close();
             context.done(); //don't return error here as we'll just end up processing the same invalid message over and over
             return;
         }
         var msg = {
             topic: "signals", 
             items: mySbMsg.content,
             entities: mySbMsg.entities,
             type: mySbMsg.type,
             generated_at: (new Date()).toISOString()
         };
         var message = new Message(JSON.stringify(msg));
         message.ack = 'full';
         message.messageId = mySbMsg.correlation_id;
         message.expiryTimeUtc = new Date((new Date()).getTime() + 1000*60*60*6); //2-day expiration
         context.log('Prepared message at ' + new Date());

         serviceClient.send(mySbMsg.device_code, message, function(err, obj){
            if (err){
                 context.log('Error sending message', err);
                 context.log('Transport-specific response', obj);            
            }
            context.log('Finished sending message at ' + new Date());
            serviceClient.close();
            context.log('Sent message for device ' + mySbMsg.device_code + ' to IoT Hub: ' + message.getData());
            context.done(err);    
         });
       }
     });
    
};
```

# Console log of the issue:
```2018-10-02T15:24:16.138 [Info] Received request at Tue Oct 02 2018 15:24:16 GMT+0000 (Coordinated Universal Time)
2018-10-02T15:24:16.138 [Info] Prepared message at Tue Oct 02 2018 15:24:16 GMT+0000 (Coordinated Universal Time)
2018-10-02T15:24:17.851 [Info] Received request at Tue Oct 02 2018 15:24:17 GMT+0000 (Coordinated Universal Time)
2018-10-02T15:24:17.851 [Info] Prepared message at Tue Oct 02 2018 15:24:17 GMT+0000 (Coordinated Universal Time)
```
There is never any confirmation that the hub receives the message, and the callback never fires. Sometimes, but not always, the following error appears

```
2018-10-02T15:22:06.614 [Error] Exception while executing function: Functions.IotHubDispatcher. mscorlib: TypeError: value.copy is not a function
    at TypeDesc.write_long [as write] (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:254:15)
    at types.Writer.write_fixed_width (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:741:14)
    at types.Writer.write_value (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:787:14)
    at types.Writer.write (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:779:14)
    at types.Writer.write_compound (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:814:18)
    at types.Writer.write_value (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:791:14)
    at types.Writer.write (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\types.js:779:14)
    at Object.message.encode (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\message.js:215:16)
    at Sender.send (D:\home\site\wwwroot\IotHubDispatcher\node_modules\rhea\lib\link.js:332:50)
    at constructor.send (D:\home\site\wwwroot\IotHubDispatcher\node_modules\azure-iot-amqp-base\lib\sender_link.js:310:72).
```

The weird thing is, this was working up until last week with v1.1.7 of the library. Then overnight it stopped working and even though we updated to the latest version of the library it did not resolve the issue.

**Comment by michaelbironneau** [comment]
Typical - spent the whole day debugging this and only after filing the issue do we notice where the problem lies: 
```
message.expiryTimeUtc = new Date((new Date()).getTime() + 1000*60*60*6);
```
I'm guessing the new version of the library wants this to be a string whereas the previous version was happy to accept a Date object. Perhaps it would be possible to have an Exception thrown if any of the message properties do not conform to the expected types, but that's a whole different issue.

**Comment by az-iot-builder-01** [comment]
@michaelbironneau, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
